### PR TITLE
[QoS] Introduce v0 of Noop QoS module

### PR DESCRIPTION
**Pull request by adshmh**

### CODE REVIEW

File: qos/noop/context.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	presetFailureResponse *HTTPResponse
+}
+
+func (rc *requestContext) GetServicePayload() protocol.Payload {
+	// TODO_TECHDEBT: log a
+	return protocol.Payload{
+		Method: http.MethodPost,
+		Data:   string(rc.httpRequestBody),
+		// TODO_TECHDEBT(@adshmh): support customization of the service request's timeout.
+		// TimeoutMillisec: time.Duration(60) * time.Second
```
Comment (Olshansk) [inline qos/noop/context.go:27]: Can we move this into a local context with TODO_TECDHEBT and set it explicitly?

File: qos/noop/context.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	}
+}
+
+func (rc *requestContext) UpdateWithResponse(endpointAddr protocol.EndpointAddr, endpointSerializedResponse []byte) {
```
Comment (Olshansk) [inline qos/noop/context.go:31]: Please add godocs. Ditto lsewhere.

File: qos/noop/selector.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+	"math/rand"
+
+	"github.com/buildwithgrove/path/protocol"
+)
+
+var _ protocol.EndpointSelector = RandomEndpointSelector{}
+
+type RandomEndpointSelector struct{}
+
+func (r RandomEndpointSelector) Select(endpoints []protocol.Endpoint) (protocol.EndpointAddr, error) {
```
Comment (Olshansk) [inline qos/noop/selector.go:14]: A few comments explaining that we select one randomly would be great!

**Review by Olshansk — CHANGES_REQUESTED** [review]
@adshmh LGTM w/ no blockers, but would appreciate it we put a bit more effort into comments, PR description,e tc...

File: qos/noop/context.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+import (
+	"net/http"
+
+	"github.com/buildwithgrove/path/gateway"
+	"github.com/buildwithgrove/path/message"
+	"github.com/buildwithgrove/path/protocol"
+)
+
+var _ gateway.RequestQoSContext = &requestContext{}
```
[the remark this answers is not in the document]
Comment (adshmh) [reply qos/noop/context.go:11]: No good reason, this was a test of how fast a NoOp QoS could be implemented.
I'll update the comments to be consistent with the rest of the code base.

File: qos/noop/context.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+type requestContext struct {
+	httpRequestBody []byte
+
+	receivedResponses []endpointResponse
+
+	presetFailureResponse *HTTPResponse
+}
+
+func (rc *requestContext) GetServicePayload() protocol.Payload {
+	// TODO_TECHDEBT: log a
```
[the remark this answers is not in the document]
Comment (adshmh) [reply qos/noop/context.go:22]: Thanks, was moved somewhere else. Fixed.

File: qos/noop/context.go [hunk 6]
```diff
@@ -8,30 +8,51 @@ import (
...
 
+	// endpointResponseTimeoutMillisec specifies the timeout for receiving a response from an endpoint serving
+	// the request represented by this requestContext instance.
+	endpointResponseTimeoutMillisec int
+
+	// receivedResponses maintains response(s) received from one or more endpoints, for the
+	// request represented by this instance of requestContext.
 	receivedResponses []endpointResponse
 
+	// presetFailureResponse, if set, is used to return a preconstructed response to the user.
```
Comment (Olshansk) [inline qos/noop/context.go:30]: this is great, ty!

**Review by Olshansk — APPROVED** [review]

File: qos/noop/http.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+	return h.payload
+}
+
+// GetHTTPStatusCode returns the HTTP status code of the user-facing HTTP response.
+// This method implements the gateway.HTTPResponse interface.
+func (h *HTTPResponse) GetHTTPStatusCode() int {
+	return h.httpStatusCode
+}
+
+// GetHTTPHeaders always returns nil, as HTTP headers are not used by noop QoS as of PR #106.
```
Comment (commoddity) [inline qos/noop/http.go:31]:
```suggestion
// GetHTTPHeaders always returns nil, as HTTP headers are not used by noop QoS as of PR #106.
// See: https://github.com/buildwithgrove/path/pull/106
```

**Review by commoddity — APPROVED** [review]
LGTM also. 

I am assuming that the actual selection logic for when to use the No-Op QoS (ie. modifying/removing the `service_alias.go` and `cmd/qos.go` files) is coming in a subsequent PR?

**Pull request merged**